middleware/kubernetes/autopath.go

```diff
@@ -0,0 +1,79 @@
+package kubernetes
+
+import "github.com/miekg/dns"
+
+// AutoPathWriter implements a ResponseWriter that also does the following:
+// * reverts question section of a packet to its original state.
+//   This is done to avoid the 'Question section mismatch:' error in client.
+// * Defers write to the client if the response code is NXDOMAIN.  This is needed
+//   to enable further search path probing if a search was not sucessful.
+// * Allow forced write to client regardless of response code.  This is needed to
+//   write the packet to the client if the final search in the path fails to
+//   produce results.
+// * Overwrites response code with AutoPathWriter.Rcode if the response code
+//   is NXDOMAIN (NameError).  This is needed to support the AutoPath.OnNXDOMAIN
+//   function, which returns a NOERROR to client instead of NXDOMAIN if the final
+//   search in the path fails to produce results.
+
+type AutoPathWriter struct {
+	dns.ResponseWriter
+	original dns.Question
+	Rcode    int
+	Sent     bool
+}
+
+// NewAutoPathWriter returns a pointer to a new AutoPathWriter
+func NewAutoPathWriter(w dns.ResponseWriter, r *dns.Msg) *AutoPathWriter {
+	return &AutoPathWriter{
+		ResponseWriter: w,
+		original:       r.Question[0],
+		Rcode:          dns.RcodeNameError,
+	}
+}
+
+// WriteMsg records the status code and calls the
+// underlying ResponseWriter's WriteMsg method unless
+// the write is deferred.
+func (r *AutoPathWriter) WriteMsg(res *dns.Msg) error {
+	if res.Rcode == dns.RcodeNameError {
+		res.Rcode = r.Rcode
+	}
+	for _, a := range res.Answer {
+		if r.original.Name == a.Header().Name {
+			continue
+		}
+		cname := dns.CNAME{
+			Hdr: dns.RR_Header{
+				Name:   r.original.Name,
+				Rrtype: dns.TypeCNAME,
+				Class:  dns.ClassINET,
+				Ttl:    a.Header().Ttl},
+			Target: dns.Fqdn(a.Header().Name)}
+		res.Answer = append(res.Answer, &cname)
```

Comment:
The way it behaves works for what we want to do... the records appended in the loop are not ranged over.